File: packages/lattice-rule-sdk-rust/src/verdict.rs
//! The verdict, and the shapes of it the type system refuses to build.
//!
//! The host calls a verdict that breaks one of its obligations **hollow** and
//! refuses it — a `fail` naming no finding, a `pass` carrying one, an `unknown`
//! with no reason, a reason on a verdict that is not `unknown`
//! (`../../lattice/src/custom-rules/host.mjs`, `verdictViolation`). Every one
//! of those is a rule that returned a shape instead of a judgment, and the
//! refusal is what stops it being read as one.
//!
//! This module's job is to make as many of them as possible impossible to write
//! rather than merely refused later:
//!
//! - [`Verdict::fail`] takes [`Findings`], which cannot be empty. There is no
//!   way to spell a failing verdict that names nothing.
//! - [`Verdict::pass`] takes nothing, so it cannot carry a finding.
//! - [`Verdict::unknown`] and [`Verdict::not_applicable`] take their reason as
//!   an argument rather than a setter, so the reason cannot be forgotten, and
//!   neither reason field is reachable from any other constructor.
//! - [`Finding::at`] takes a file and a position together, and each coordinate
//!   as a [`NonZeroU32`], so a position with no file and a zeroth line are both
//!   unwritable.
//!
//! What the type system cannot hold is emptiness of a `String`: an empty
//! message or an empty reason is still refused by the host, and this crate does
//! not check it a second time — see the crate docs on not duplicating the
//! host's validation.

use std::num::NonZeroU32;

use serde::Serialize;

use crate::CONTRACT;

/// The four-state vocabulary every judgment in this engine speaks.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum VerdictKind {
    /// The rule looked and found nothing.
    Pass,
    /// The rule found something, and every finding names what.
    Fail,
    /// The rule could not reach a judgment, and says why.
    Unknown,
    /// The rule does not apply to this run, and says why.
    NotApplicable,
}

/// One thing a rule found.
///
/// `id` must be an id this rule's own catalogue declares — the host refuses a
/// finding it cannot resolve to a catalogue entry, because SARIF would drop it.
/// That check is the host's and is deliberately not repeated here.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Finding {
    id: String,
    message: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    source_file: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    line: Option<NonZeroU32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    column: Option<NonZeroU32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    project: Option<String>,
}

impl Finding {
    /// A finding with no location, naming the catalogue entry it belongs to and
    /// what it found.
    ///
    /// The message is what a reader acts on, so it is the finding's own
    /// sentence rather than the catalogue's template — the catalogue says what
    /// the id MEANS, this says what happened here.
    #[must_use]
    pub fn new(id: impl Into<String>, message: impl Into<String>) -> Self {
        Self {
            id: id.into(),
            message: message.into(),
            source_file: None,
            line: None,
            column: None,
            project: None,
        }
    }

    /// Places the finding in a file, with no position inside it.
    ///
    /// The case this exists for is a graph edge: an edge may carry a
    /// `sourceFile` and never carries a line, so a rule judging edges can name
    /// the file honestly and must not invent `1:1` to fill the shape.
    #[must_use]
    pub fn in_file(mut self, source_file: impl Into<String>) -> Self {
        self.source_file = Some(source_file.into());
        self
    }

    /// Places the finding at a 1-based position in a file.
    ///
    /// File and position move together because the host refuses a position with
    /// no file — "a position with no file sends a reader to a line in nothing" —
    /// and [`NonZeroU32`] is why a zeroth line cannot be written at all.
    /// [`crate::ImportRecord::position`] hands over the pair an import site
    /// already carries.
    #[must_use]
    pub fn at(
        mut self,
        source_file: impl Into<String>,
        line: NonZeroU32,
        column: NonZeroU32,
    ) -> Self {
        self.source_file = Some(source_file.into());
        self.line = Some(line);
        self.column = Some(column);
        self
    }

    /// Attributes the finding to a project.
    #[must_use]
    pub fn in_project(mut self, project: impl Into<String>) -> Self {
        self.project = Some(project.into());
        self
    }
}

/// One or more findings.
///
/// A `Vec<Finding>` that is guaranteed not to be empty, which is the whole
/// point: it is the argument type of [`Verdict::fail`], so "this rule failed
/// and named nothing" has no spelling. There is no `len`/`is_empty` pair on
/// purpose — the answer to `is_empty` is always `false`, and a caller asking it
/// is a caller who has forgotten what this type is for.
#[derive(Debug, Clone)]
pub struct Findings {
    items: Vec<Finding>,
}

impl Findings {
    /// Starts a non-empty list from the finding that makes it non-empty.
    #[must_use]
    pub fn new(first: Finding) -> Self {
        Self { items: vec![first] }
    }

    /// Adds a finding, keeping the list non-empty by construction.
    pub fn push(&mut self, finding: Finding) {
        self.items.push(finding);
    }

    /// Adds a finding, for building a list in one expression.
    #[must_use]
    pub fn with(mut self, finding: Finding) -> Self {
        self.items.push(finding);
        self
    }

    /// The non-empty list a `Vec` holds, or `None` when it holds nothing.
    ///
    /// `None` is the caller's cue that there is nothing to fail over — and
    /// [`Verdict::from_findings`] is the one-line way to answer it correctly.
    #[must_use]
    pub fn from_vec(findings: Vec<Finding>) -> Option<Self> {
        if findings.is_empty() {
            None
        } else {
            Some(Self { items: findings })
        }
    }

    /// The findings, in the order they were added.
    #[must_use]
    pub fn as_slice(&self) -> &[Finding] {
        &self.items
    }
}

/// What a rule answers.
///
/// Constructed only through the four verdict constructors, so the invariants
/// the host checks hold by construction rather than by care.
#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Verdict {
    contract: u32,
    verdict: VerdictKind,
    /// Never skipped, even when empty: "every verdict states its findings, and
    /// a clean rule states the empty list rather than leaving the field out".
    findings: Vec<Finding>,
    #[serde(skip_serializing_if = "Option::is_none")]
    reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    not_applicable_reason: Option<String>,
}

impl Verdict {
    /// The rule looked and found nothing.
    ///
    /// This is a CLAIM about the workspace, not a shrug: it says the rule read
    /// the evidence it declared it needs and there was nothing to report. A
    /// rule that could not read that evidence answers [`Verdict::unknown`]
    /// instead (`../../../AGENTS.md`).
    #[must_use]
    pub fn pass() -> Self {
        Self::of(VerdictKind::Pass, Vec::new())
    }

    /// The rule found something, and the findings say what.
    #[must_use]
    pub fn fail(findings: Findings) -> Self {
        Self::of(VerdictKind::Fail, findings.items)
    }

    /// The rule could not reach a judgment, and `reason` says why.
    ///
    /// This is the verdict for every path that would otherwise return an empty
    /// list it did not earn: a parameter that is not there, an evidence kind
    /// that arrived malformed, two halves of the bundle that disagree. Any
    /// `unknown` makes the run exit 3 — "could not look" is a distinct outcome
    /// from "looked and found nothing", and the exit codes keep it that way.
    #[must_use]
    pub fn unknown(reason: impl Into<String>) -> Self {
        Self {
            reason: Some(reason.into()),
            ..Self::of(VerdictKind::Unknown, Vec::new())
        }
    }

    /// The rule does not apply to this run, and `reason` says why.
    ///
    /// Not a quieter `pass`: it is counted and reported separately, because a
    /// rule that never applies is a rule nobody is being protected by, and that
    /// has to be visible rather than absorbed into a passing count.
    #[must_use]
    pub fn not_applicable(reason: impl Into<String>) -> Self {
        Self {
            not_applicable_reason: Some(reason.into()),
            ..Self::of(VerdictKind::NotApplicable, Vec::new())
        }
    }

    /// `pass` when the list is empty, `fail` when it is not.
    ///
    /// The shape most rules end in, and the one place emptiness legitimately
    /// means "clean" — because the caller has just finished looking. A rule that
    /// reaches here without having looked has already gone wrong somewhere
    /// above, which is what [`Verdict::unknown`] is for.
    #[must_use]
    pub fn from_findings(findings: Vec<Finding>) -> Self {
        match Findings::from_vec(findings) {
            Some(findings) => Self::fail(findings),
            None => Self::pass(),
        }
    }

    /// Which of the four this is.
    #[must_use]
    pub fn kind(&self) -> VerdictKind {
        self.verdict
    }

    /// The verdict as the UTF-8 JSON text the ABI carries back.
    ///
    /// # Panics
    ///
    /// Never in practice: a verdict holds strings and small integers, which
    /// always serialize. A panic here reaches the host as a trap it names,
    /// which is the loud direction rather than a verdict invented on the way
    /// out.
    #[must_use]
    pub fn to_json(&self) -> String {
        serde_json::to_string(self)
            .expect("a verdict is strings and integers, which always serialize")
    }

    fn of(verdict: VerdictKind, findings: Vec<Finding>) -> Self {
        Self {
            contract: CONTRACT,
            verdict,
            findings,
            reason: None,
            not_applicable_reason: None,
        }
    }
}

#[cfg(test)]
mod tests {
    use std::num::NonZeroU32;

    use super::{Finding, Findings, Verdict, VerdictKind};

    fn one() -> NonZeroU32 {
        NonZeroU32::new(1).expect("1 is not zero")
    }

    #[test]
    fn a_pass_states_the_contract_and_an_empty_findings_list() {
        assert_eq!(
            Verdict::pass().to_json(),
            r#"{"contract":1,"verdict":"pass","findings":[]}"#
        );
    }

    #[test]
    fn a_fail_carries_every_optional_key_the_host_allows() {
        let verdict = Verdict::fail(Findings::new(
            Finding::new("cross-tag-edge", "beta reaches gamma")
                .at(
                    "packages/beta/src/lib.rs",
                    one(),
                    NonZeroU32::new(5).expect("5 is not zero"),
                )
                .in_project("beta"),
        ));
        assert_eq!(
            verdict.to_json(),
            r#"{"contract":1,"verdict":"fail","findings":[{"id":"cross-tag-edge","message":"beta reaches gamma","sourceFile":"packages/beta/src/lib.rs","line":1,"column":5,"project":"beta"}]}"#
        );
    }

    #[test]
    fn a_finding_with_no_position_omits_the_position_keys() {
        let verdict = Verdict::fail(Findings::new(
            Finding::new("cross-tag-edge", "beta reaches gamma")
                .in_file("packages/beta/Cargo.toml"),
        ));
        assert_eq!(
            verdict.to_json(),
            r#"{"contract":1,"verdict":"fail","findings":[{"id":"cross-tag-edge","message":"beta reaches gamma","sourceFile":"packages/beta/Cargo.toml"}]}"#
        );
    }

    #[test]
    fn unknown_and_not_applicable_each_carry_their_own_reason_and_no_other() {
        assert_eq!(
            Verdict::unknown("params.forbiddenTag is not a string").to_json(),
            r#"{"contract":1,"verdict":"unknown","findings":[],"reason":"params.forbiddenTag is not a string"}"#
        );
        assert_eq!(
            Verdict::not_applicable("no project carries the tag").to_json(),
            r#"{"contract":1,"verdict":"not_applicable","findings":[],"notApplicableReason":"no project carries the tag"}"#
        );
    }

    // The silent direction: an empty finding list must never serialize as a
    // failing verdict, and a rule that found something must never serialize as
    // a passing one. `Verdict::fail` cannot be called with an empty list at
    // all — that is the type system's half — so what is left to pin is the
    // collapse `from_findings` performs.
    #[test]
    fn from_findings_passes_on_nothing_and_fails_on_something() {
        assert_eq!(Verdict::from_findings(Vec::new()).kind(), VerdictKind::Pass);
        assert_eq!(
            Verdict::from_findings(vec![Finding::new("cross-tag-edge", "beta reaches gamma")])
                .kind(),
            VerdictKind::Fail
        );
        assert!(Findings::from_vec(Vec::new()).is_none());
    }

    #[test]
    fn findings_keep_the_order_they_were_added() {
        let mut findings = Findings::new(Finding::new("first-finding", "one"));
        findings.push(Finding::new("second-finding", "two"));
        let findings = findings.with(Finding::new("third-finding", "three"));
        let messages: Vec<&str> = findings
            .as_slice()
            .iter()
            .map(|finding| finding.message.as_str())
            .collect();
        assert_eq!(messages, ["one", "two", "three"]);
    }
}

File: packages/archkeep/src/governance/verdict.mjs
/**
 * The canonical 4-state verdict vocabulary every governance capability emits
 * and consumes — the single answer shape an architecture-governance envelope
 * carries.
 *
 * The vocabulary exists because the three statuses the envelope has always
 * used (`"ok"` / `"findings"` / `"no-verdict"`) are about THIS run's position
 * on THIS tree, while the governance wave needs a verdict an evidencing
 * process can compare across runs, feeds, and capabilities. The mapping is
 * the interesting part and it is one-way on purpose:
 *
 *   `"ok"`         → `"pass"`
 *   `"findings"`   → `"fail"`
 *   `"no-verdict"` → `"unknown"`
 *
 * and `"not_applicable"` has no source status in this release — no envelope
 * status maps to it, so it never rides an envelope. Engine behavior does emit
 * it today, on the paths that cannot judge: the fitness registry answers it
 * for a declared function whose `match` selects no observed project, and
 * again for `coverage-minimum` on a path-scoped run, which no scoped run can
 * judge; decision fitness derives it for a decision without authority; the
 * boundary metrics answer it when there is no boundary config to judge
 * against (`./metrics.mjs`). A
 * custom rule may return it as its own verdict — the host requires a
 * `notApplicableReason` (`../custom-rules/host.mjs`) — but no engine path
 * answers it on a rule's behalf.
 *
 * ## The invariants, I1–I5
 *
 * I1. `pass` implies complete coverage — a run that could not fully read the
 *     tree can never pass. A `pass` without coverage is a claim the evidence
 *     cannot back.
 * I2. `fail` implies at least one finding — a failing verdict that names no
 *     finding leaves the reader to guess what failed.
 * I3. `unknown` implies "could not look" — coverage incomplete, an unresolved
 *     question, or a thrown analysis. `unknown` is never a degraded `pass`.
 * I4. `not_applicable` requires `notApplicableReason` — the reader has to be
 *     able to tell WHY a rule did not apply, since "did not apply" and "did
 *     not run" are indistinguishable otherwise.
 * I5. The cardinal rule: an analysis that failed, or a rule that could not
 *     determine, must emit `unknown`, NEVER `pass`. `pass` is the loudest
 *     claim the vocabulary makes and the hardest to disprove, so every other
 *     state exists to refuse it.
 *
 * The enforcer that makes the invariants executable is `buildDecision`, in
 * this file, beside the vocabulary it enforces — a vocabulary and the check
 * that a verdict's evidence agrees with it are one subject, and splitting
 * them across the report boundary had made the core verdict module
 * (`../verdict.mjs`) depend on the presentation layer. `../report/evidence.mjs`
 * re-exports `buildDecision` so the render-side callers keep their import
 * path — a path, never a second implementation.
 */

import { describe, isNonEmptyString } from "../values.mjs";

/** The four canonical verdict values. */
export const VERDICTS = Object.freeze(["pass", "fail", "unknown", "not_applicable"]);

/** The single mapping from an envelope status to a verdict. */
export const VERDICT_FOR_STATUS = Object.freeze({
  // used by its own test
  ok: "pass",
  findings: "fail",
  "no-verdict": "unknown",
});

/**
 * Whether a string names one of the four verdicts.
 *
 * @param {unknown} value
 * @returns {boolean}
 */
export function isVerdict(value) {
  return typeof value === "string" && VERDICTS.includes(value);
}

/**
 * The verdict an envelope status implies. Unknown statuses refuse loudly
 * rather than map to a guess — a status this vocabulary has never heard of is
 * a caller reading from the future, and guessing which verdict it implied
 * would make the evidence lie about what actually ran.
 *
 * @param {string} status One of `"ok"` | `"findings"` | `"no-verdict"`.
 * @returns {"pass"|"fail"|"unknown"}
 * @throws {Error} when `status` is not one of the three the envelope can hold.
 */
export function verdictForStatus(status) {
  const verdict = VERDICT_FOR_STATUS[status];
  if (verdict === undefined) {
    throw new Error(
      `archkeep: status "${status}" has no verdict — expected one of ` +
        `${Object.keys(VERDICT_FOR_STATUS).join(", ")}`,
    );
  }
  return verdict;
}

/**
 * One function's verdict record — the evidence envelope every governance
 * consumer (a fitness function, a waiver judge) reads.
 *
 * `evidence` is an object of deterministic facts the verdict is a claim over
 * (the same facts a report row renders), and `message` is human text naming
 * what was decided and why. `rows` is optional per-function observed detail
 * (matched projects, judged edges); it rides along so a report can show the
 * function's coverage without re-deriving it a second way.
 *
 * A `not_applicable` verdict carries its `notApplicableReason` (invariant I4):
 * "did not apply" and "did not run" are indistinguishable otherwise, and a
 * reader has to be told which one this was.
 *
 * @param {{verdict: (typeof VERDICTS)[number], name: string, evidence: object,
 *   message: string, rows?: object[], notApplicableReason?: string}} decision
 * @returns {object}
 * @throws {Error} on a verdict outside the four — a programming error in the
 *   function that built the decision, not a fact about the workspace — and on
 *   a `not_applicable` verdict without its `notApplicableReason`.
 */
export function fitnessVerdict({ verdict, name, evidence, message, rows, notApplicableReason }) {
  if (!isVerdict(verdict)) {
    throw new Error(
      `archkeep: fitness function "${name}" returned verdict ${JSON.stringify(verdict)} — ` +
        `expected one of ${VERDICTS.join(", ")}.`,
    );
  }
  if (verdict === "not_applicable" && !isNonEmptyString(notApplicableReason)) {
    throw new Error(
      `archkeep: fitness function "${name}" returned "not_applicable" without ` +
        `notApplicableReason — invariant I4: the reader must be told why the ` +
        `function did not apply.`,
    );
  }
  return {
    verdict,
    name,
    evidence,
    message,
    ...(rows === undefined ? {} : { rows }),
    ...(notApplicableReason === undefined ? {} : { notApplicableReason }),
  };
}

/**
 * The decision builder: turns a command's verdict counts into the `decision`
 * the envelope optionally carries, enforcing the five invariants the module
 * header states in code rather than leaving them to a docs page a later
 * command author might not read.
 *
 * This function decides nothing about whether a finding IS one — the command
 * that built the envelope owns that. What it decides is whether the verdict
 * and its evidence AGREE, and it throws when they do not, the same posture
 * `../report/json.mjs`'s `jsonEnvelope` takes for the three consistency rules
 * it enforces: a mismatch here is a bug in the command, not a fact about the
 * workspace.
 *
 * The shape it produces:
 *
 *   {
 *     verdict: "pass" | "fail" | "unknown" | "not_applicable",
 *     reason?: string,                // always present for unknown
 *     notApplicableReason?: string,   // always present for not_applicable
 *     sampleTime?: string             // opt-in, never on a deterministic envelope
 *   }
 *
 * A caller may pass `reason` for `unknown` — it names WHICH could-not-look
 * condition fired (coverage incomplete, an unresolved intent boundary, a
 * thrown analysis). Without it, `buildDecision` states the generic one. The
 * reason field itself is always present on an `unknown` decision (I3).
 *
 * ## Determinism is the default
 *
 * The envelope this decision rides on is byte-deterministic
 * (`docs/reference/json-output.md`: no timestamp, no random identifier). So
 * `sampleTime` is OPT-IN by construction: a command passes it explicitly when
 * it is an age/count capability (waivers, debt, health — the features
 * `./clock.mjs` serves), and a command whose verdict must stay reproducible
 * over an unchanged tree emits a decision with no time at all. That is how
 * the determinism↔time tension is resolved — the clock is injectable (a test
 * drives the same code with a fixed time), never asserted from the wall
 * clock.
 *
 * ## One refusal per invariant
 *
 * I1 refuses a `pass` over incomplete coverage — the same refusal
 * `jsonEnvelope` makes for `status: "ok"` over incomplete coverage, at the
 * verdict layer — and a `pass` carrying findings. I2 refuses a `fail` that
 * names no finding. I3 keeps a reason on every `unknown`, defaulting the
 * generic one when the caller supplies none. I4 refuses a `not_applicable`
 * without its `notApplicableReason`. I5 is I1's first check plus every caller
 * choosing `unknown` wherever the run did not reach a verdict.
 *
 * `not_applicable` has no envelope status, so `buildDecision` reaches it only
 * through an explicit `verdict` — the route a Fitness or Waiver capability
 * takes. Engine behavior today never passes it: `jsonEnvelope` refuses a
 * `decision.verdict` that contradicts the envelope's `status`, and no status
 * maps to `not_applicable`, so the state is locked out of every envelope this
 * release builds.
 *
 * @param {{
 *   verdict?: "pass"|"fail"|"unknown"|"not_applicable",
 *   status?: "ok"|"findings"|"no-verdict",
 *   coverageComplete: boolean,
 *   findings: number,
 *   reason?: string|null,
 *   notApplicableReason?: string|null,
 *   sampleTime?: string
 * }} run
 * @returns {{verdict: string, reason?: string, notApplicableReason?: string,
 *   sampleTime?: string}}
 * @throws {Error} on any invariant violation (I1–I4).
 */
export function buildDecision(run) {
  if (run.verdict === undefined && run.status === undefined) {
    // No status, no explicit verdict — a builder called with neither is a
    // programming error, not a fact about the workspace.
    throw new Error("archkeep: buildDecision needs either a status or an explicit verdict");
  }
  const verdict = run.verdict ?? verdictForStatus(run.status);
  // The vocabulary latch. `verdictForStatus` cannot produce a stranger, so
  // this refuses exactly the explicitly-passed verdicts — and it must run
  // BEFORE the status-conflict check below, so a malformed verdict reports
  // the vocabulary it violated rather than a conflict it never had. Without
  // this latch, "perhaps", "PASS", `42` and `{}` fell through the
  // pass/fail/unknown arms into the not_applicable one and SHIPPED whenever
  // `notApplicableReason` happened to be truthy — a byte-legal decision
  // about a workspace nobody judged.
  if (!isVerdict(verdict)) {
    throw new Error(
      `archkeep: refusing to build a decision with verdict ${describe(run.verdict)} — ` +
        `expected one of ${VERDICTS.join(", ")}. This is a bug in the command that ` +
        `built the decision.`,
    );
  }
  if (
    run.verdict !== undefined &&
    run.status !== undefined &&
    run.verdict !== verdictForStatus(run.status)
  ) {
    throw new Error(
      `archkeep: refusing to build a decision where verdict "${run.verdict}" contradicts status ` +
        `"${run.status}" — status implies ${verdictForStatus(run.status)}, and a decision that ` +
        `disagrees with its own status would make one of the two a lie. ` +
        `This is a bug in the command that built the decision.`,
    );
  }

  // The `findings` count is the cardinal evidence number — a non-negative
  // integer that the I1–I5 invariants all rely on. A missing, non-numeric, or
  // negative value would silently falsify every comparison (`undefined > 0` is
  // `false`), producing a clean verdict over a run whose counts were never set
  // or are logically impossible — the exact silent direction this module exists
  // to refuse.
  if (typeof run.findings !== "number" || !Number.isFinite(run.findings) || run.findings < 0) {
    throw new Error(
      `archkeep: refusing to build a decision where findings is ${JSON.stringify(run.findings)} ` +
        `— findings must be a non-negative number, or the verdict invariants cannot be enforced. ` +
        `This is a bug in the command that built the decision.`,
    );
  }
  if (verdict === "pass") {
    if (run.coverageComplete !== true) {
      throw new Error(
        `archkeep: refusing to emit a "pass" decision over incomplete coverage ` +
          `(coverage.complete: ${run.coverageComplete}) — a run that could not fully read the ` +
          `tree can never pass. This is a bug in the command that built the decision.`,
      );
    }
    if (run.findings > 0) {
      throw new Error(
        `archkeep: refusing to emit a "pass" decision with ${run.findings} finding(s) — ` +
          `"pass" and "fail" cannot both be true of the same run. This is a bug in the command.`,
      );
    }
    return withSampleTime({ verdict }, run.sampleTime);
  }

  if (verdict === "fail") {
    if (run.findings < 1) {
      throw new Error(
        `archkeep: refusing to emit a "fail" decision with no findings — a failing verdict ` +
          `must name what failed. This is a bug in the command that built the decision.`,
      );
    }
    return withSampleTime({ verdict }, run.sampleTime);
  }

  if (verdict === "unknown") {
    // An absent reason (undefined/null) still defaults to the generic one —
    // absence is not emptiness. A SUPPLIED reason must actually say
    // something: "" and "   " are byte-present but semantically absent, and
    // a non-string reason would ship a `typeof` artifact where the reader
    // was promised a sentence (I3).
    const reason =
      run.reason ??
      (run.coverageComplete === true ? "no verdict was reached" : "coverage was incomplete");
    if (!isNonEmptyString(reason)) {
      throw new Error(
        `archkeep: refusing to emit an "unknown" decision with a ${describe(run.reason)} reason — ` +
          `I3 requires a non-empty reason naming why no verdict was reached. ` +
          `This is a bug in the command that built the decision.`,
      );
    }
    return withSampleTime({ verdict, reason }, run.sampleTime);
  }

  // verdict === "not_applicable" (I4). The reason must be a string a reader
  // could act on — the falsy check this replaces refused "" but SHIPPED
  // "   " and non-strings, the same byte-present-semantically-absent hole.
  if (!isNonEmptyString(run.notApplicableReason)) {
    throw new Error(
      `archkeep: refusing to emit a "not_applicable" decision without notApplicableReason — ` +
        `"did not apply" and "did not run" must never be indistinguishable, and a reason of ` +
        `${describe(run.notApplicableReason)} tells the reader nothing. ` +
        `This is a bug in the command that built the decision.`,
    );
  }
  return withSampleTime({ verdict, notApplicableReason: run.notApplicableReason }, run.sampleTime);
}

/**
 * Adds `sampleTime` to the decision only when the caller opted into time —
 * the determinism rule in `buildDecision`'s header. Absent `sampleTime`, the
 * decision object carries exactly the invariant-bearing fields and nothing
 * more.
 *
 * @param {object} decision
 * @param {string|undefined} sampleTime
 * @returns {object}
 */
function withSampleTime(decision, sampleTime) {
  return sampleTime === undefined ? decision : { ...decision, sampleTime };
}
